### PR TITLE
Changes to `Document` 

### PR DESCRIPTION
**Pull request by sandreae**
Some additions to `DocumentView` which I feel are needed when working with this data type in `aquadoggo`. Make `DocumentView` on `Document` optional in order to represent deleted documents.

## Reasoning
- more separation wanted between document view fields / values and operation types
- a document which has been deleted generates a new document view Id but no actual view. We need a way of representing this state in `Document`.
- in `aquadoggo` we need to know which operation transported the current value on a documents fields in order to store it in the db. This is somewhat due to the current db layout, however as I thought about it more I feel it's nice metadata to have anyway (for debugging, and sparce materialisation). It also requires little extra processing overhead as we can easily add the operation id's when resolving the document view. 

### `DocumentViewFields` 
the fields of a document view as `String` -> `DocumentViewValue` mappings
### `DocumentViewValue`
struct containing the `OperationId` and `OperationValue` for a field on a document view
### `Document` changes
Make `DocumentView` optional

closes: #316 

## 📋 Checklist

- [x] Add tests that cover your changes
- [x] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [x] Link this PR to any issues it closes
- [x] New files contain a SPDX license header
- [x] Check if descriptions and terminology match `handbook` content (and visa-versa)

### CODE REVIEW

File: p2panda-rs/src/document/mod.rs
Comment (cafca) [inline p2panda-rs/src/document/mod.rs:192]: :D

Comment (sandreae) [reply p2panda-rs/src/document/mod.rs:192]: :p

**Comment by sandreae** [comment]
Hmmmm.....

Now I'm not completely sure about this pattern, it ends up in a strange pattern of having to insert deleted `DocumentViews` into the database, and if we want to reconstruct them then we need to persist the `is_deleted` status of each document view (also a weird pattern, if a document/document view is deleted, it doesn't exist anymore). 

I think this maybe just highlights that we (or just me?) don't have a really clear idea of how we want to handle deletion in the database. Any ideas :rofl:

**Comment by sandreae** [comment]
The above comment doesn't change anything about needing field ids and the other changes here, just how/where/when/what etc.. we do about deletion wrt documents and document views.

**Comment by sandreae** [comment]
I think the answer is that document views don't really know about deletion... `Documents` do, and if they are deleted, then we decide what to do at an implementation level. When a document gets deleted no new document view is created.

**Comment by sandreae** [comment]
We could make `DocumentView` on `Document` optional here:

https://github.com/p2panda/p2panda/blob/19a6612affc5875953e13d7cb32ea89edbb09587/p2panda-rs/src/document/document.rs#L82-L88

**Comment by sandreae** [comment]
I'm :+1: this myself, I think it's a nice solution.

File: p2panda-rs/src/document/document.rs
Comment (cafca) [inline p2panda-rs/src/document/document.rs:49]: I just noticed a problem with this `IsEdited` logic. It will return the `IsEdited` marker for documents that were created and immediately deleted even though such a document has never been edited. I think we should initialise this as `false` and then set it to `true` when encountering the first edit instead.

Comment (cafca) [reply p2panda-rs/src/document/document.rs:49]: This would also be a good thing to put into the docstring.

**Review by cafca — CHANGES_REQUESTED** [review]
The changes are good! You could add a sentence about the subtleties of document deletion to the handbook https://p2panda.org/handbook/docs/organising-data/documents#document-views

File: p2panda-rs/src/document/document.rs
Comment (sandreae) [reply p2panda-rs/src/document/document.rs]: Yes, nice, good spot!

Here are the handbook changes: https://github.com/p2panda/handbook/pull/170

**Comment by cafca** [comment]
Is there something wrong with codecoverage again? I noticed the coverage changes, for example in `definitions.rs` but don't understand why those tests aren't run anymore?

https://app.codecov.io/gh/p2panda/p2panda/compare/319/changes

**Comment by sandreae** [comment]
Yeh, codecov seems weird, I'm sure I got the coverage into positive too...

File: p2panda-rs/src/document/document.rs
Comment (adzialocha) [inline p2panda-rs/src/document/document.rs:47]: The backticks in the docstring are a little off 😧

File: p2panda-rs/src/document/document_view_fields.rs
Comment (adzialocha) [inline p2panda-rs/src/document/document_view_fields.rs:24]: Newline missing

**Review by adzialocha — COMMENTED** [review]
Sorry for commenting on these little things, they are mostly notes to myself (or someone is faster than me). I can change them myself later as soon as I'm back at the computer.

**Comment by sandreae** [comment]
Regarding code-cov, it seems to have been this commit where the coverage went wonky: https://app.codecov.io/gh/p2panda/p2panda/commit/8d6185d1506a753b3dd0d8c575071f285104ca99

**Comment by sandreae** [comment]
And that commit only changed doc strings....

**Comment by sandreae** [comment]
I think this is just a glitch, maybe it doesn't block merging?

**Comment by sandreae** [comment]
codecov sorted after merging main.

**Pull request merged**
